hash-tables/problem-set-hash-tables.md

```diff
@@ -0,0 +1,287 @@
+# Problem Set: Hash Table
+
+## Directions
+
+Complete all questions below. **Solve all questions using a hash table** (Python dictionary).
+
+## Practice
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: 9U2j0J
+* title: Hash Tables
+* points: 3
+### !question
+
+Write a function `get_intersection` that takes two integer lists with unique values and returns their intersection in a new array.
```

Comment:
Don't get me wrong, I really like the problem, but just note that Python does have built in set operations like intersection.

```python
def get_intersection(red, blue):
    red_set = set(red)
    blue_set = set(blue)
    return list(red_set.intersection(blue_set))
```

Reply:
The following direction says not to use anything like this. Is the concern that students will anyway and that we'll need to remember to manually review the automated test results? Or is it more that we're possibly showing students a suboptimal approach?

If the former, we could at least make it harder to use set by replacing the set ctor to make it unusable (though I don't think we can easily prevent the use of {} literals). Something like

```python
import main as m

def set(*args):
    msg = "set class unavailable for these tests"
    raise NotImplementedError(msg)
m.set = set
```

This produces an error in the test as follows, which I think is very legible and description. 😁 
```
======================================================================
ERROR: test_get_intersection_many_matches (test_methods.TestChallenge)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/app/test_methods.py", line 31, in test_get_intersection_many_matches
    result = get_intersection(red, blue)
  File "/usr/src/app/main.py", line 2, in get_intersection
    red_set = set(red)
  File "/usr/src/app/test_methods.py", line 12, in set
    raise NotImplementedError(msg)
NotImplementedError: set class unavailable for these tests
```

If the latter, we could more explicitly acknowledge the existence of the set class as a better way of approaching this, but that we're explicitly practicing the use of dicts here.